cvmfs/s3fanout.cc

```diff
@@ -244,29 +245,53 @@ void *S3FanoutManager::MainUpload(void *data) {
   LogCvmfs(kLogS3Fanout, kLogDebug, "Upload I/O thread started");
   S3FanoutManager *s3fanout_mgr = static_cast<S3FanoutManager *>(data);
 
+  s3fanout_mgr->InitPipeWatchFds();
+
   // Don't schedule more jobs into the multi handle than the maximum number of
   // parallel connections.  This should prevent starvation and thus a timeout
   // of the authorization header (CVM-1339).
   unsigned jobs_in_flight = 0;
 
-  while (s3fanout_mgr->thread_upload_run_) {
-    JobInfo *info = NULL;
-    {
-      MutexLockGuard m(s3fanout_mgr->jobs_todo_lock_);
-      if (!s3fanout_mgr->jobs_todo_.empty() &&
-          (jobs_in_flight < s3fanout_mgr->pool_max_handles_)) {
-        info = s3fanout_mgr->jobs_todo_.back();
-        s3fanout_mgr->jobs_todo_.pop_back();
+  while (true) {
+    // Check events with 100ms default timeout or timeout suggested by curl
+    int timeout = 100;
+    // Cannot use int64_t because curl_multi_timeout header is platform-specific
+    long curl_timeout;  // NOLINT
+    curl_multi_timeout(s3fanout_mgr->curl_multi_, &curl_timeout);
```

Comment:
I think we have to go another route.  As per the documentation, we should not use this function together with the multi socket API (which we use).